feature/add GetQuestion endpoint

Levliv opened the pull request:
Created GetQuestionCommand.cs, QuestionResponse.cs (QuestionResponseMapper.cs), OptionInfo.cs (OptionInfoMapper.cs), UserAnswerInfo.cs (UserAnswerInfoMapper.cs),
QuestionController.cs edited

Levliv committed:
Created GetQuestionCommand.cs, QuestionResponse.cs (QuestionResponseMapper.cs), OptionInfo.cs (OptionInfoMapper.cs), UserAnswerInfo.cs (UserAnswerInfoMapper.cs),
QuestionController.cs edited

Levliv committed:
Fixed AutoInject DI

Levliv committed:
Fixed style QuestionResponseMapper.cs

Levliv committed:
Fixed style in QuestionController.cs

AlfaRedDog commented on src/SurveyService.Models.Dto/Responses/Question/QuestionResponse.cs:
```diff
@@ -0,0 +1,21 @@
+  public bool IsObligatory { get; set; }
+  public bool IsPrivate { get; set; }
+  public bool HasMultipleChoice { get; set; }
+  public bool HasCustomOptions { get; set; }
```
Maybe you should put the question properties in the QuestionInfo model? Get Group also uses this model

vorobevamv replied:
@AlfaRedDog not exactly the same: in Response there is IsActive; in Info - it shouldn't be (only active question are included); may be some more properties are unnecessary in Info (think about it)

Levliv committed:
Minor logic changes in Mappers, QuestionRepository.cs - in progress

Levliv committed:
Prep for futher rabbit implement pub/sub pat

Levliv committed:
Fixed minor mistakes in GetQuestionCommand.cs

Levliv committed:
GetQuestionCommand.cs - fixed minor mistakes
IUserInfoMapper.cs - fixed DI
IUserService.cs - fixed style
QuestionRepository.cs - changes some controversial line
RabbitMqConfig.cs - added GetUsersDataEndpoint - to get user info from UserService
UserService.cs - Fixed typing

Levliv committed:
Fixed error in CreateUserAnswerRequestValidator.cs,
Added IncludeUserInfo filter to GetQuestionCommand.cs,
Fixed style OptionInfoMapper.cs,
Added IncludeAnswers filter to QuestionRepository.cs,
Deleted UserId from UserAnswerInfo.cs (as it is in UserInfo model),
Deleted UserId from UserAnswerInfoMapper.cs (as deleted from UserInfoMapper)

Levliv committed:
Fixed style

maybee16 commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,83 @@
+
+    DbQuestion dbQuestion = await _repository.GetAsync(filter);
+    List<OptionInfo> optionInfos = new();
+    foreach (DbOption option in dbQuestion.Options)
```
Add empty line

Levliv replied:
Done

maybee16 commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,83 @@
+    foreach (DbOption option in dbQuestion.Options)
+    {
+      List<UserAnswerInfo> userAnswerInfos = new();
+      foreach (DbUserAnswer optionUsersAnswer in option.UsersAnswers)
```
Add empty line

Levliv replied:
Done

maybee16 commented on src/SurveyService.Data/QuestionRepository.cs:
```diff
@@ -35,6 +37,37 @@ public Task<DbQuestion> GetAsync(Guid questionId)
+    
+    IQueryable<DbQuestion> dbQuestions = _provider.Questions.AsQueryable();
+    dbQuestions = dbQuestions.Where(question => question.Id == filter.QuestionId);
+    if (filter.IncludeAnswers)
```
Add empty line

Levliv replied:
Done

Levliv committed:
Fixed style in GetQuestionCommand.cs,
Minor changes in GetQuestionFilter.cs

vorobevamv commented on src/SurveyService/Controllers/QuestionController.cs:
```diff
@@ -18,4 +20,12 @@ public class QuestionController : ControllerBase
\ No newline at end of file
+  
+  [HttpGet("get")]
+  public async Task<OperationResultResponse<QuestionResponse>> CreateAsync(
```
Create - ?

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Models.Dto/Requests/Question/Filters/GetQuestionFilter.cs:
```diff
@@ -0,0 +1,19 @@
+public class GetQuestionFilter
+{
+  [FromQuery(Name = "questionid")]
+  public Guid? QuestionId { get; set; }
```
not nullable! it's get endpoint - Id must be in request

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Models.Dto/Requests/Question/Filters/GetQuestionFilter.cs:
```diff
@@ -0,0 +1,19 @@
+  public Guid? QuestionId { get; set; }
+
+  [FromQuery(Name = "includecustomoptions")]
+  public bool IncludeCustomOptions { get; set; } = true;
```
includeOptions (not custom)
case: user needs just the question content - includeOptions=false (other includes are "false" too);
case: user needs question and options to peek  - includeOptions=true (return only not custom; other includes are "false");
case: user wants to see results - includeOptions=true, IncludeAnswers = true (return all options + answers Ids)
and so on

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Models.Dto/Requests/Question/Filters/GetQuestionFilter.cs:
```diff
@@ -0,0 +1,19 @@
+  public bool IncludeCustomOptions { get; set; } = true;
+  
+  [FromQuery(Name = "includeanswers")]
+  public bool IncludeAnswers { get; set; }
```
may be set it 'false' by default?

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Models.Dto/Requests/Question/Filters/GetQuestionFilter.cs:
```diff
@@ -0,0 +1,19 @@
+  public bool IncludeAnswers { get; set; }
+  
+  [FromQuery(Name = "includeuserinfo")]
+  public bool IncludeUserInfo { get; set; }
```
may be set it 'false' by default too?

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,84 @@
+  
+  public async Task<OperationResultResponse<QuestionResponse>> ExecuteAsync(GetQuestionFilter filter)
+  {
+    OperationResultResponse<QuestionResponse> response = new();
```
what for you create it here? you need response only at the end of command, move it there;
and make
OperationResultResponse<QuestionResponse> response = new(body: ....);

Levliv replied:
Added check for dbQuestion.IsAnonymous.
For now if question IsAnonymous I return questions without userInfo and add corresponding error message.

vorobevamv commented on src/SurveyService.Validation/UserAnswer/CreateUserAnswerRequestValidator.cs:
```diff
@@ -51,14 +46,10 @@ public class CreateUserAnswerRequestValidator : AbstractValidator<(CreateUserAns
-            .Select(q => q.Id).ToList().Except(x.dbOptions.Select(o => o.QuestionId).Distinct().ToList()).Any())
-        .WithMessage("Not all required questions from the group were answered.");
+        .WithMessage("Trying to answer twice in a question.");
     });
```
do not remove!

Levliv replied:
Ok)

AlfaRedDog commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,84 @@
+  {
+    OperationResultResponse<QuestionResponse> response = new();
+
+    if (filter is null || filter.QuestionId is null)
```
QuestionId is required, you don't have to check it

Levliv replied:
Done

AlfaRedDog reviewed (commented):
I may have missed it, but you don't check deadline, hasRealTimeResult, IsAnonymous, IsPrivate properties when you take the survey results

AlfaRedDog commented on src/SurveyService.Mappers/Models/Interfaces/IOptionInfoMapper.cs:
```diff
@@ -0,0 +1,13 @@
+using LT.DigitalOffice.Kernel.Attributes;
+using LT.DigitalOffice.Models.Broker.Models;
```
remove unused

Levliv replied:
Done

AlfaRedDog commented on src/SurveyService.Mappers/Models/OptionInfoMapper.cs:
```diff
@@ -0,0 +1,30 @@
+  public OptionInfoMapper(
+    IUserAnswerInfoMapper userAnswerInfoMapper)
+  {
+    _mapper = userAnswerInfoMapper;
```
remove unused mapper

Levliv replied:
Done

Levliv replied:
Done

AlfaRedDog commented on src/SurveyService.Mappers/Responses/QuestionResponseMapper.cs:
```diff
@@ -0,0 +1,42 @@
+    IOptionInfoMapper optionInfoMapper
+    )
+  {
+    _mapper = optionInfoMapper;
```
remove unused mapper

Levliv replied:
Done

Vasis3038 commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,84 @@
+  private readonly IResponseCreator _responseCreator;
+  private readonly IUserService _userService;
+  private readonly IOptionInfoMapper _optionInfoMapper;
+  private readonly IUserInfoMapper _userInfoMapper;
```
Do you need it here?

Levliv replied:
Why not?

Vasis3038 replied:
You don't use it

Levliv replied:
IUserService - line 95
IOptionINfoMapper - line 101
IUserInfoMapper - line 97

Levliv replied:
If you are about IUserInfoMapper - fixed

Vasis3038 commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,84 @@
+    IResponseCreator responseCreator,
+    IUserService userService,
+    IOptionInfoMapper optionInfoMapper,
+    IUserInfoMapper userInfoMapper,
```
Do you need it here?

Levliv replied:
Done

Vasis3038 commented on src/SurveyService.Broker/Requests/UserService.cs:
```diff
@@ -0,0 +1,50 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.SurveyService.Broker.Requests
```
Use file-scoped namespace

Levliv replied:
Done

Levliv committed:
Fixed CreateGetPredicates in QuestionRepository.cs

Levliv committed:
Fixed _mapper in QuestionResponseMapper.cs,

Levliv committed:
Fixed checks in GetQuestionCommand.cs

Levliv committed:
Changed questionid to required

Levliv committed:
Added checks in GetQuestionCommand.cs,
added includeOptions in GetQuestionFilter.cs,
Fixed QuestionResponseMapper.cs

Levliv committed:
Fixed QuestionController.cs

Levliv committed:
Fixed using in IOptionInfoMapper.cs

Levliv committed:
Fixed file-scoped namespace in UserService.cs

Levliv committed:
Merge remote-tracking branch 'origin/develop' into feature/GetQuestion_endpoint

# Conflicts:
#	src/SurveyService.Validation/UserAnswer/CreateUserAnswerRequestValidator.cs

Levliv committed:
Fixed minor mistakes in GetQuestionCommand.cs, OptionInfoMapper.cs

Levliv committed:
Fixed unused usings in UserInfo.cs

vorobevamv commented on src/SurveyService.Models.Dto/Requests/Question/Filters/GetQuestionFilter.cs:
```diff
@@ -0,0 +1,24 @@
+public class GetQuestionFilter
+{
+  [Required]
+  [FromQuery(Name = "questionid")]
```
don't need [Required] here
Guid is value type, it can't be null

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Models.Dto/Requests/Question/Filters/GetQuestionFilter.cs:
```diff
@@ -0,0 +1,24 @@
+  [FromQuery(Name = "includeoptions")]
+  public bool IncludeOptions { get; set; } = false;
+  
+  [FromQuery(Name = "includecustomoptions")]
```
it is unnecessary
IncludeOptions=true + IncludeAnswers=false => IncludeCustomOptions=false
IncludeAnswers=true => IncludeCustomOptions=true

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,114 @@
+  
+  public async Task<OperationResultResponse<QuestionResponse>> ExecuteAsync(GetQuestionFilter filter)
+  {
+
```
remove empty line

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,114 @@
+  {
+
+    OperationResultResponse<QuestionResponse> response = new();
+    if (filter is null)
```
can filter be null at all?
please, check. i think it should be caught in pipeline

Levliv replied:
Do you mean in ASP.NET Core Middleware?

vorobevamv replied:
yes

Levliv replied:
Fixed

vorobevamv commented on src/SurveyService.Data/QuestionRepository.cs:
```diff
@@ -35,6 +62,15 @@ public Task<DbQuestion> GetAsync(Guid questionId)
 
+  public Task<DbQuestion> GetAsync(GetQuestionFilter filter)
+  {
+    return filter is null
```
can filter be null?

Levliv replied:
As it is API, maybe someone will use this method from another place and pass null deliberately

vorobevamv replied:
OK. It can be. it's unlikely but possible

vorobevamv commented on src/SurveyService.Data/QuestionRepository.cs:
```diff
@@ -11,6 +13,31 @@ public class QuestionRepository : IQuestionRepository
+
+    return dbQuestions;
+  }
   public QuestionRepository(
```
add empty line

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Data/QuestionRepository.cs:
```diff
@@ -11,6 +13,31 @@ public class QuestionRepository : IQuestionRepository
+  private IQueryable<DbQuestion> CreateGetPredicates(
+    GetQuestionFilter filter,
+    IQueryable<DbQuestion> dbQuestions)
+  {
```
may be add check if dbQuestions.Any()?
to not to check filter predicates, if there is no Question
//or even may be make this check in Get method - think about it

Levliv replied:
Should we? Result will stay the same, but we add some useless lines of code

vorobevamv replied:
lines of code vs checking empty collection...
may be. OK, leave it without Any - check

vorobevamv commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,114 @@
+    }
+    else
+    {
+      response.Body = _questionResponseMapper.Map(dbQuestion);
```
use in command only questionResponseMapper; call optionInfoMapper in questionResponseMapper, userAnswerInfoMapper in questionResponseMapper and so on

Levliv replied:
Well, is what I did right?

vorobevamv replied:
Sorry, my bad. It's all-right about mapping in common. But what for you create OptionInfo, UserInfo.. in command? You need just dbQuestion and usersData

Levliv replied:
Done in waterfall way

vorobevamv commented on src/SurveyService.Models.Dto/Requests/Question/Filters/GetQuestionFilter.cs:
```diff
@@ -0,0 +1,24 @@
+  public Guid QuestionId { get; set; }
+
+  [FromQuery(Name = "includeoptions")]
+  public bool IncludeOptions { get; set; } = false;
```
set it "true" by default

Levliv replied:
Done

Levliv committed:
action

Levliv committed:
Fixed style

Levliv committed:
Fixed mapping logic in GetQuestionCommand.cs

Levliv committed:
Style changes

Levliv committed:
Errors in rabbitmq

Vasis3038 commented on src/SurveyService.Data/QuestionRepository.cs:
```diff
@@ -11,6 +13,25 @@ public class QuestionRepository : IQuestionRepository
+    {
+      dbQuestions = dbQuestions
+        .Include(question => question.Options.Where(option => option.IsActive))
+        .ThenInclude(option => option.UsersAnswers.Where(ua => filter.IncludeAnswers));
```
Do you need "Where" here?
(ua => filter.IncludeAnswers) true for each UserAnswer..

Levliv replied:
Done

Levliv committed:
Fixed style

VLiamin commented on src/SurveyService.Broker/Requests/UserService.cs:
```diff
@@ -0,0 +1,48 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.SurveyService.Broker.Requests;
```
empty line

Levliv replied:
Done

VLiamin commented on src/SurveyService.Broker/Requests/UserService.cs:
```diff
@@ -0,0 +1,48 @@
+  public async Task<List<UserData>> GetUsersDataAsync(
+    List<Guid> usersIds,
+    List<string> errors,
+    CancellationToken cancellationToken = default
```
parenthesis

Levliv replied:
Done

VLiamin commented on src/SurveyService.Broker/Requests/UserService.cs:
```diff
@@ -0,0 +1,48 @@
+
+    object request = IGetUsersDataRequest.CreateObj(usersIds);
+    
+    List<UserData> usersData = (await _rcGetUsersData.ProcessRequest<IGetUsersDataRequest, IGetUsersDataResponse>(
```
May be remove variable?

Levliv replied:
Done

VLiamin commented on src/SurveyService.Broker/Requests/UserService.cs:
```diff
@@ -0,0 +1,48 @@
+      return null;
+    }
+
+    object request = IGetUsersDataRequest.CreateObj(usersIds);
```
May be remove variable?

Levliv replied:
Done

VLiamin commented on src/SurveyService.Data/Interfaces/IQuestionRepository.cs:
```diff
@@ -13,6 +14,7 @@ public interface IQuestionRepository
   Task<bool> CheckGroupProperties(Guid groupId, DateTime? deadline, bool hasRealTimeResult);
 
   Task<DbQuestion> GetAsync(Guid questionId);
+  Task<DbQuestion> GetAsync(GetQuestionFilter filter);
```
empty line

Levliv replied:
Done

VLiamin commented on src/SurveyService.Data/QuestionRepository.cs:
```diff
@@ -11,6 +13,25 @@ public class QuestionRepository : IQuestionRepository
+    {
+      dbQuestions = dbQuestions
+        .Include(question => question.Options.Where(option => option.IsActive))
+          .ThenInclude(option => option.UsersAnswers);
```
May be not do this tab?

Levliv replied:
Done

VLiamin commented on src/SurveyService.Mappers/Responses/QuestionResponseMapper.cs:
```diff
@@ -0,0 +1,52 @@
+        HasMultipleChoice = dbQuestion.HasMultipleChoice,
+        HasCustomOptions = dbQuestion.HasCustomOptions,
+        IsActive = dbQuestion.IsActive,
+        Options = optionInfos?
```
What is Infos mean? May be info?

Levliv replied:
Renamed to optionsInfo - contains the information about all the options for the question

vorobevamv commented on src/SurveyService.Data/QuestionRepository.cs:
```diff
@@ -11,6 +13,25 @@ public class QuestionRepository : IQuestionRepository
+        .Include(question => question.Options.Where(option => option.IsActive))
+          .ThenInclude(option => option.UsersAnswers);
+    }
+    else
```
just ELSE?
includeOptions can be true and can be false
it can be three cases:
options=true + answers=true
options=false + answers=false
options=true + answers=false
(options=false + answers=true is not valid, ignore it and return options=false + answers=false)

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Data/QuestionRepository.cs:
```diff
@@ -11,6 +13,25 @@ public class QuestionRepository : IQuestionRepository
+    else
+    {
+      dbQuestions = dbQuestions
+        .Include(question => question.Options.Where(option => option.IsActive));
```
if includeOptions=true and includeAnswers=false include only active and **not-custom** options

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Data/QuestionRepository.cs:
```diff
@@ -11,6 +13,25 @@ public class QuestionRepository : IQuestionRepository
 
+  private IQueryable<DbQuestion> CreateGetPredicates(
+    GetQuestionFilter filter,
+    IQueryable<DbQuestion> dbQuestions)
```
you get question by Id -> it can be question (not list) or null in result

Levliv replied:
It wouldn't be a list as id is unique, but can be null. Check added

vorobevamv commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,106 @@
+        new List<string> { "You must enter 'questionid'." });
+    }
+
+    DbQuestion dbQuestion = await _repository.GetAsync(filter);
```
add check: if dbQuestion is null, return NotFound

Levliv replied:
Done

SolovyovDmitry commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,106 @@
+        {
+          List<string> errors = new();
+          UserData userData = filter.IncludeUserInfo
+            ? (await _userService.GetUsersDataAsync(new List<Guid> { optionUsersAnswer.UserId }, errors)).FirstOrDefault()
```
It's certainly would be a better solution to collect all the required users and then send them once instead of sending each one every time. So time of communication between services will be reduced significantly

Levliv replied:
Fixed

SolovyovDmitry commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,106 @@
+      List<string> totalErrors = new ();
+      foreach (DbOption option in dbQuestion.Options)
+      {
+        List<Tuple<DbUserAnswer, UserData>> usersInfo = new();
```
Is it possible to use dictionary instead of list of tuples?

Levliv replied:
Recombined mappers structure, so now it's just List<UserData>

Levliv committed:
Fixed mapping logic

Levliv committed:
Fixed checks in GetQuestionCommand.cs

Levliv committed:
Fixed UserService.cs

Levliv committed:
Fixed UserService.cs

Levliv committed:
Fixed style

Levliv committed:
Fixed style in QuestionRepository.cs

Levliv committed:
Fixed b logic

Levliv committed:
Fixed mistake with request processing

Levliv committed:
Style

Levliv committed:
Style

vorobevamv commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,94 @@
+      .ToList();
+    List<string> errors = new();
+    List<UserData> usersData = filter.IncludeUserInfo
+      ? await _userService.GetUsersDataAsync(users, errors)
```
lines 75-77 remove
line 80 to
      ? await _userService.GetUsersDataAsync(
        dbQuestion.Options.SelectMany(x => x.UsersAnswers.Select(y => y.UserId)).Distinct().ToList(),
        errors)

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,94 @@
+      ? await _userService.GetUsersDataAsync(users, errors)
+      : null;
+    
+    if (errors.Count != 0)
```
I think, don't need it; just return this errors in common response

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Data/QuestionRepository.cs:
```diff
@@ -11,6 +13,32 @@ public class QuestionRepository : IQuestionRepository
+    }
+    else
+    {
+      dbQuestions = dbQuestions?.Where(question => question.Id == filter.QuestionId);
```
what is it for here? you've already filtered by QuestionId in Get

Levliv replied:
Done

vorobevamv commented on src/SurveyService.Mappers/Responses/QuestionResponseMapper.cs:
```diff
@@ -0,0 +1,52 @@
+
+    foreach (DbOption dbQuestionOption in dbQuestion.Options)
+    {
+      Console.Write(dbQuestionOption.Content);
```
what is it for?

Levliv replied:
Removed

vorobevamv commented on src/SurveyService.Mappers/Responses/QuestionResponseMapper.cs:
```diff
@@ -0,0 +1,52 @@
+public class QuestionResponseMapper : IQuestionResponseMapper
+{
+  private readonly IOptionInfoMapper _optionInfoMapper;
+  private readonly IUserAnswerInfoMapper _userAnswerInfoMapper;
```
remove

Levliv replied:
Done

Levliv committed:
Fixed some mistakes

Levliv committed:
Fixed style

Levliv committed:
Merge branch 'develop' into feature/GetQuestion_endpoint

Levliv committed:
Merge branch 'develop' into feature/GetQuestion_endpoint

Levliv committed:
Update UserInfoMapper.cs

Levliv committed:
Fixed minor mistakes

Levliv committed:
Fixed error in body in GetQuestionCommand.cs

maybee16 commented on src/SurveyService.Business/Commands/Question/GetQuestionCommand.cs:
```diff
@@ -0,0 +1,89 @@
+        HttpStatusCode.NotFound,
+        new List<string> {"Question with specified id wasn't found"});
+    }
+    if (dbQuestion.Deadline > DateTime.Now && !dbQuestion.HasRealTimeResult)
```
Add space

Levliv committed:
Minor style changes

Levliv committed:
Merge branch 'develop' into feature/GetQuestion_endpoint
